src/algo/isomorphism.rs

```diff
@@ -631,8 +667,94 @@ mod matching {
                     }
                 }
             }
+            if result.is_some() {
+                return result;
+            }
+        }
+        result
+    }
+
+    pub struct GraphMatcher<'a, 'b, 'c, G0, G1, NM, EM>
+    where
+        G0: NodeCompactIndexable
+            + EdgeCount
+            + GetAdjacencyMatrix
+            + GraphProp
+            + IntoNeighborsDirected,
+        G1: NodeCompactIndexable
+            + EdgeCount
+            + GetAdjacencyMatrix
+            + GraphProp
+            + IntoNeighborsDirected,
+        NM: NodeMatcher<G0, G1>,
+        EM: EdgeMatcher<G0, G1>,
+    {
+        st: (Vf2State<'a, G0>, Vf2State<'b, G1>),
+        node_match: &'c mut NM,
+        edge_match: &'c mut EM,
+        match_subgraph: bool,
+        stack: Vec<Frame<G0, G1>>,
+    }
+
+    impl<'a, 'b, 'c, G0, G1, NM, EM> GraphMatcher<'a, 'b, 'c, G0, G1, NM, EM>
+    where
+        G0: NodeCompactIndexable
+            + EdgeCount
+            + GetAdjacencyMatrix
+            + GraphProp
+            + IntoNeighborsDirected,
+        G1: NodeCompactIndexable
+            + EdgeCount
+            + GetAdjacencyMatrix
+            + GraphProp
+            + IntoNeighborsDirected,
+        NM: NodeMatcher<G0, G1>,
+        EM: EdgeMatcher<G0, G1>,
+    {
+        pub fn new(
+            g0: &'a G0,
+            g1: &'b G1,
+            node_match: &'c mut NM,
+            edge_match: &'c mut EM,
+            match_subgraph: bool,
+        ) -> Self {
+            let stack = vec![Frame::Outer];
+            Self {
+                st: (Vf2State::new(g0), Vf2State::new(g1)),
+                node_match,
+                edge_match,
+                match_subgraph,
+                stack,
+            }
+        }
+    }
```

Comment:
Add a `size_hint` upper bound when the graphs are small enough.

Reply:
Are you referring to [size_hint method](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.size_hint)?
Hmm. I don't think this is deterministic as it depends on the graph. The worst case scenario where every permutation is an isomorphism has a time complexity of O(n! * n) according to [this paper](https://www.researchgate.net/publication/224279054_Performance_evaluation_of_the_VF_graph_matching_algorithm). So should the upper bound for the size_hint be `n! * n` ? If so, should we return None if n is too large (when calculating `n!` is too expensive)?

This is a screenshot of the relevant image in that article:
![image](https://user-images.githubusercontent.com/2531700/176515081-d7ff1903-0887-452d-b30a-e04f5e2133d4.png)

Reply:
Yes. It's mostly interesting for graphs with just a couple nodes, but you can encode the values that fit in an usize, for n≤19.

Reply:
You are right. Time complexity is different from the actual number of solutions. I said that because I couldn't find a quick reference to talk about it. The formula you suggest calculates the number of permutations which is definitely the worst case when n=m.